[JSC] Optimize 16-bit string comparison in DFG and FTL

sosukesuzuki opened the pull request:
#### e4b576d5067fea02f30320e1ad3541678db46725
<pre>
[JSC] Optimize 16-bit string comparison in DFG and FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=295374">https://bugs.webkit.org/show_bug.cgi?id=295374</a>

Previously, JSC&apos;s DFG and FTL JIT compilers only generated fast machine code for
string equality comparison when both operands were 8-bit strings. Other cases,
including 16-bit string comparison, fell back to the slow path (operationCompareStringEq).

This patch extends the optimization to handle 16-bit string comparisons as well.
When both operands are 16-bit strings, we now generate an inline comparison loop
similar to the 8-bit case, avoiding the expensive C++ call.  Mixed encoding comparisons
continue to use the slow path.

                                        TipOfTree                  Patched

string-equality-mixed-encoding        2.6945+-0.3466            2.5821+-0.0914          might be 1.0435x faster
string-equality-8bit-8bit             4.8215+-0.7379            4.8174+-0.9536
string-equality-long-16bit            2.5633+-0.4814            2.1642+-0.4516          might be 1.1844x faster
string-equality                       4.4529+-0.1344     ?      4.7678+-0.2863        ? might be 1.0707x slower
string-equality-16bit-16bit           6.8205+-0.3002     ^      3.9166+-0.4730        ^ definitely 1.7414x faster

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

* JSTests/microbenchmarks/string-equality-16bit-16bit.js: Added.
(shouldBe):
* JSTests/microbenchmarks/string-equality-8bit-8bit.js: Added.
(shouldBe):
* JSTests/microbenchmarks/string-equality-long-16bit.js: Added.
(shouldBe):
* JSTests/microbenchmarks/string-equality-mixed-encoding.js: Added.
(shouldBe):
* JSTests/stress/string-equality-16bit-16bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-16bit-8bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-8bit-16bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-8bit-8bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-ascii-16bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-emoji.js: Added.
(shouldBe):
* JSTests/stress/string-equality-japanese-characters.js: Added.
(shouldBe):
* JSTests/stress/string-equality-null-characters.js: Added.
(shouldBe):
* JSTests/stress/string-equality-surrogate-pairs.js: Added.
(shouldBe):
* JSTests/stress/string-equality-unicode-scripts.js: Added.
(shouldBe):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e4b576d5067fea02f30320e1ad3541678db46725

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116611 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60852 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38832 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84094 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113533 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24693 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99578 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64535 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24056 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17718 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60406 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103076 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94078 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17777 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119401 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109139 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37625 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27940 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93060 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95849 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92882 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37894 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15638 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33599 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37520 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42991 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133414 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37183 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36050 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38891 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -20064,32 +20067,66 @@ IGNORE_CLANG_WARNINGS_END
         LValue leftData = m_out.loadPtr(left, m_heaps.StringImpl_data);
         LValue rightData = m_out.loadPtr(right, m_heaps.StringImpl_data);
 
-        ValueFromBlock indexAtStart = m_out.anchor(length);
```
Maybe we can put this inside a block, so we don't have to rename these variables?

sosukesuzuki replied:
Wrapped loops inside blocks!

sosukesuzuki committed:
[JSC] Optimize 16-bit string comparison in DFG and FTL
https://bugs.webkit.org/show_bug.cgi?id=295374

Previously, JSC's DFG and FTL JIT compilers only generated fast machine code for
string equality comparison when both operands were 8-bit strings. Other cases,
including 16-bit string comparison, fell back to the slow path (operationCompareStringEq).

This patch extends the optimization to handle 16-bit string comparisons as well.
When both operands are 16-bit strings, we now generate an inline comparison loop
similar to the 8-bit case, avoiding the expensive C++ call.  Mixed encoding comparisons
continue to use the slow path.

                                        TipOfTree                  Patched

string-equality-mixed-encoding        2.6945+-0.3466            2.5821+-0.0914          might be 1.0435x faster
string-equality-8bit-8bit             4.8215+-0.7379            4.8174+-0.9536
string-equality-long-16bit            2.5633+-0.4814            2.1642+-0.4516          might be 1.1844x faster
string-equality                       4.4529+-0.1344     ?      4.7678+-0.2863        ? might be 1.0707x slower
string-equality-16bit-16bit           6.8205+-0.3002     ^      3.9166+-0.4730        ^ definitely 1.7414x faster

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

* JSTests/microbenchmarks/string-equality-16bit-16bit.js: Added.
(shouldBe):
* JSTests/microbenchmarks/string-equality-8bit-8bit.js: Added.
(shouldBe):
* JSTests/microbenchmarks/string-equality-long-16bit.js: Added.
(shouldBe):
* JSTests/microbenchmarks/string-equality-mixed-encoding.js: Added.
(shouldBe):
* JSTests/stress/string-equality-16bit-16bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-16bit-8bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-8bit-16bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-8bit-8bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-ascii-16bit.js: Added.
(shouldBe):
* JSTests/stress/string-equality-emoji.js: Added.
(shouldBe):
* JSTests/stress/string-equality-japanese-characters.js: Added.
(shouldBe):
* JSTests/stress/string-equality-null-characters.js: Added.
(shouldBe):
* JSTests/stress/string-equality-surrogate-pairs.js: Added.
(shouldBe):
* JSTests/stress/string-equality-unicode-scripts.js: Added.
(shouldBe):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
